### ISSUE TITLE
command_on_download should wait for tagging

### ISSUE THREAD

**Issue by boyska** [gpodder/gpodder#1485]
I use both `tagging` and `command_on_download`. Unfortunately, the two are not running in the intended order, so the files are added to VLC playlist without the new metadata, which makes tagging effectively useless.

Can we make tagging happen earlier? I haven't found an obvious way to determine the order in which extensions hook will run.

**Comment by auouymous** [comment]
The extensions are loaded in the same order as the directory listing. You could try disabling the extension, closing gpodder, prefixing the `command_on_download.py` file with a `z` to sort it last, restart gpodder and enable the extension. This will only work if python is sorting the directory listing, otherwise the file will still be in the same unsorted location. If that is the case, you would need to copy `command_on_download.py` to `zcommand_on_download.py` and then delete `command_on_download.py`.

Gpodder could iterate the `extensions.enabled` setting to load them. You could then disable an extension and re-enable it to move it to last place. Of course that would need to be done whenever a new extension is enabled because it would then become last.

**Comment by boyska** [comment]
>  prefixing the command_on_download.py file with a z to sort it last, restart gpodder and enable the extension

I could do that, to check if that's fixing the problem. Would you be willing to enable this style of sorting upstream? I'm ok doing that, but I really don't want to maintain my own fork :)

**Comment by boyska** [comment]
Yes, it works. I used numbers as a prefix, which I found more "typical". tagging became `20_tagging.py` while `command_on_download` became `80_command_on_download`.

I think every other extension should become prefixed with `50_`, so that ordering is more explicit.

**Comment by elelay** [comment]
yes, it's a common solution for ordering script invocation in debian installer etc.

**Comment by auouymous** [comment]
Oh right, a numeric prefix could be ignored by gpodder and prevent existing extensions from being disabled when the user upgrades to this new system. So `20_tagging.py` is stored in `extensions.enabled` as just `tagging`.

Gpodder would need to perform a numeric sort unless python guarantees the directory listing will be sorted. User extensions in `~/Gpodder/Extensions/` would need to be sorted into the built-in list as well. We can prefix all built-in extensions but should user extensions without a prefix be sorted first, last or as 50? Also, enabling an extension must use a sorted insert when adding it to the list, otherwise it will be in the wrong order until the user restarts gpodder.

**Comment by boyska** [comment]
>  We can prefix all built-in extensions but should user extensions without a prefix be sorted first, last or as 50? 

my understanding is that if an extension developer never cared about sorting, they can go in whatever position. Apparently I'm the first person to care, so I guess that's not a big issue.

by the way: I think this hack about renaming files is good enough, because it's simple and it solves the most practical need.
However, if sorting hooks is a very common need, there needs to be a more complex mechanism. What if my extension need to react early on a hook, but late on another? Sorting should ideally be decided when registering to a hook.

**Comment by auouymous** [comment]
Adding a priority to each hook would be better, but is not easily possible. Each hook is a method in the extension's class. Gpodder iterates over all extensions and calls the method for all enabled extensions. Maybe a decorator could set a priority for the method, but Gpodder would need sort the extension list based on these priorities every time it calls a hook.

**Comment by dconrad5** [comment]
Just popping in to say I found the renaming of command_on_download.py to z_command_on_download.py to work as intended to have it occur after tagging and rename operations. 1 more vote of support for intentionally prefixing numbers to make the order of operations predictable 👍. Thanks for the workaround too!

**Comment by dconrad5** [comment]
For consideration: PR #1698
